install_config/registry/deploy_registry_existing_clusters.adoc

```diff
@@ -91,7 +94,147 @@ Daemonsets ensure that when nodes are created, they contain copies of a
+xref:../../dev_guide/daemonsets.adoc#dev-guide-daemonsets[Using DaemonSets].
+
+[[redeploy-registry]]
+=== Re-deploying the Registry
+
+There is a couple of things to keep on mind when re-deploying an existing docker-registry
+DeploymentConfig or DaemonSet:
+
+1. Service IP shall be preserved.
+2. Any change to the service IP is not propagated to the {product-title} cluster's components
+until the master API and controllers are restarted.
+3. The docker-registry service must exist before the docker-registry pod is created.
+
+**Service IP shall be preserved** because master caches the docker-registry's internal address.
+Unless one of the following applies, the service IP and port will be used to construct docker
+image pull-spec of the newly pushed images.
+
+- `OPENSHIFT_DEFAULT_REGISTRY` environment variable not containing
+  the service IP is passed to the master process 
+- xref:extended_registry_configuration.adoc#setting-the-registry-hostname[
+  `imagePolicyConfig:internalRegistryHostname`] not containing service IP is set in the master
+  configuration file
+
+It's important to keep the pull-spec's registry part of managed images and image stream
+repositories the same for all the images in the cluster because some tools (e.g. `oc adm prune
+images`) determine the address of the docker registry from the existing images unless directly
+provided.
+
+It's recommended to use the DNS registry name instead of the service address in any case. But even
+though used, preserving the registry's service IP will ensure consistency with whatever
+image-related resource already exists in the cluster which will make possible future migration
+easier.
+
+Before reading further, keep in mind also the following recommendations:
+
+1. Back up the existing configuration.
+2. It's usually safer and faster to modify the existing working deployment than starting from a
+scratch.
+3. When the above is satisfied, you can return back to the earlier working version with
+`oc rollout undo ...`
+
+[NOTE]
+====
+Unfortunately the `oc rollout` is currently limited just to the DeploymentConfig. Do the backup
+for the DaemonSet installation.
+====
+
+==== Determining registry's internal hostname
+
+The internal hostname is used for example to update imagestream's `status.dockerImageRepository`
+whenever it's created or updated. It's set on each new image's `dockerImageReference`. And is used
+in
+xref:../../admin_guide/image_policy.adoc#image-policy-configuring-the-image-policy-admission-plug-in[admission]
+and xref:../../admin_guide/image_policy.adoc#configuring-registries-allowed-for-import[registry
+whitelisting].
+
+To determine whether the master is configured to use the
+xref:extended_registry_configuration.adoc#setting-the-registry-hostname[DNS name] of IP address,
+execute the following in arbitrary namespace:
+
+----
+$ oc create imagestream tmpsvcip && \
+  oc get -o jsonpath=$'{.status.dockerImageRepository}\n' \
```

Comment:
AFAIK image streams are created with an IP address and it's asynchronously replaced by the FQDN, so FQDN may appear with a little delay.

Reply:
I think that happens only when you have master config misconfigured, at least from the issues we have been seeing with double deployments.

Reply:
The hostname specified in the configuration file takes precedence. If not present, `OPENSHIFT_DEFAULT_REGISTRY` env var is used. When unset, the master [evaluates](https://github.com/openshift/origin/blob/10927c956187c547bc64700198842f1daa1fdf28/pkg/cmd/server/origin/admission/plugin_initializer.go#L114) service's `"${DOCKER_REGISTRY_SERVICE_HOST}:${DOCKER_REGISTRY_SERVICE_PORT}"`.

With `oc cluster up`, all is unset until the service exists (or maybe is up - need to verify). Nevetheless, good point - I need to explain the empty output as well.

Reply:
TL;DR; if there's no output, no explicit address is configured and the service does not yet exist.